crates/uv-interpreter/src/discovery.rs

```diff
@@ -0,0 +1,1134 @@
+            0,
+            buf.as_mut_ptr().cast(),
+            buf.len() as u32 * 2,
+            &mut bytes_returned,
+            std::ptr::null_mut(),
+        ) != 0
+    };
+
+    // SAFETY: The handle is valid.
+    #[allow(unsafe_code)]
+    unsafe {
+        CloseHandle(reparse_handle);
+    }
+
+    // If the operation failed, assume it's not a reparse point.
+    if !success {
+        return false;
+    }
+
+    let reparse_point = String::from_utf16_lossy(&buf[..bytes_returned as usize]);
+    reparse_point.contains("\\AppInstallerPythonRedirector.exe")
+}
+
+/// On Unix, we do not need to deal with Windows store shims.
+///
+/// See the Windows implementation for details.
+#[cfg(not(windows))]
+fn is_windows_store_shim(_path: &Path) -> bool {
+    false
+}
+
+impl InterpreterRequest {
+    /// Create a request from a string.
+    ///
+    /// This cannot fail, which means weird inputs will be parsed as [`InterpreterRequest::File`] or [`InterpreterRequest::ExecutableName`].
+    pub fn parse(value: &str) -> Self {
+        // e.g. `3.12.1`
+        if let Ok(version) = VersionRequest::from_str(value) {
+            return Self::Version(version);
+        }
+        // e.g. `python3.12.1`
+        if let Some(remainder) = value.strip_prefix("python") {
+            if let Ok(version) = VersionRequest::from_str(remainder) {
+                return Self::Version(version);
+            }
+        }
+        // e.g. `pypy@3.12`
+        if let Some((first, second)) = value.split_once('@') {
+            if let Ok(implementation) = ImplementationName::from_str(first) {
+                if let Ok(version) = VersionRequest::from_str(second) {
+                    return Self::ImplementationVersion(implementation, version);
+                }
+            }
+        }
+        for implementation in ImplementationName::iter() {
+            if let Some(remainder) = value
+                .to_ascii_lowercase()
+                .strip_prefix(implementation.as_str())
+            {
+                // e.g. `pypy`
+                if remainder.is_empty() {
+                    return Self::Implementation(*implementation);
+                }
+                // e.g. `pypy3.12`
+                if let Ok(version) = VersionRequest::from_str(remainder) {
+                    return Self::ImplementationVersion(*implementation, version);
+                }
+            }
+        }
+        let value_as_path = PathBuf::from(value);
+        // e.g. ./path/to/.venv
+        if value_as_path.is_dir() {
+            return Self::Directory(value_as_path);
+        }
+        // e.g. ./path/to/python3.exe
+        // If it contains a path separator, we'll treat it as a full path even if it does not exist
+        if value.contains(std::path::MAIN_SEPARATOR) {
+            return Self::File(value_as_path);
+        }
+        // Finally, we'll treat it as the name of an executable (i.e. in the search PATH)
+        // e.g. foo.exe
```

Comment:
So we _do_ allow (e.g.) `--python foo` where `foo` is an arbitrary executable in `PATH`? I think the post said we wouldn't do this: https://github.com/astral-sh/uv/issues/2386#issuecomment-1992133003. Although I think we _should_ support it and don't agree with the linked comment.

Reply:
Yeah I forgot I said we shouldn't do this in that latter comment. I think it's okay? I'll consider this a bit more.. it seems nice not to have a breaking change